File: 37. Sudoku Solver.py
# 37. Sudoku Solver
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        rows = defaultdict(set)
        cols = defaultdict(set)
        boxs = defaultdict(set)
        x = []

        for i in range(9):
            for j in range(9):
                if board[i][j] != ".":
                    n = board[i][j]
                    rows[i].add(n)
                    cols[j].add(n)
                    boxs[(i//3)*3 + (j//3)].add(n)
                else:
                    x.append((i, j))
        
        def check(i, j, n):
            if n in rows[i] or n in cols[j] or n in boxs[(i//3)*3 + (j//3)]:
                return False
            return True
        
        def solve(ind):
            if ind == len(x):
                return True

            i, j = x[ind]
            boxInd = (i//3)*3 + (j//3)
            for n in "123456789":
                if check(i, j, n):
                    board[i][j] = n
                    rows[i].add(n)
                    cols[j].add(n)
                    boxs[boxInd].add(n)

                    if solve(ind + 1):
                        return True
                    
                    board[i][j] = "."
                    rows[i].remove(n)
                    cols[j].remove(n)
                    boxs[boxInd].remove(n)
        
        solve(0)

File: 37. Sudoku Solver.cpp
// 37. Sudoku Solver

#include<vector>

using namespace std;



class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        // Initialize bitmasks for rows, columns, and boxes
        vector<int> rows(9, 0), cols(9, 0), boxes(9, 0);
        
        // Fill the bitmasks based on the initial board
        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] != '.') {
                    int num = board[i][j] - '1';
                    int boxIdx = (i / 3) * 3 + (j / 3);
                    rows[i] |= (1 << num);
                    cols[j] |= (1 << num);
                    boxes[boxIdx] |= (1 << num);
                }
            }
        }
        
        solve(board, rows, cols, boxes);
    }
    
    bool solve(vector<vector<char>>& board, vector<int>& rows, vector<int>& cols, vector<int>& boxes) {
        // Find the cell with the fewest possible numbers
        int minOptions = 10;
        int row = -1, col = -1;
        
        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] == '.') {
                    int boxIdx = (i / 3) * 3 + (j / 3);
                    int used = rows[i] | cols[j] | boxes[boxIdx];
                    int options = 9 - __builtin_popcount(used);
                    
                    if (options < minOptions) {
                        minOptions = options;
                        row = i;
                        col = j;
                    }
                }
            }
        }
        
        // If no empty cells left, the board is solved
        if (minOptions == 10) {
            return true;
        }
        
        int boxIdx = (row / 3) * 3 + (col / 3);
        int used = rows[row] | cols[col] | boxes[boxIdx];
        
        // Try each possible number
        for (int num = 0; num < 9; ++num) {
            if (!(used & (1 << num))) {
                board[row][col] = '1' + num;
                rows[row] |= (1 << num);
                cols[col] |= (1 << num);
                boxes[boxIdx] |= (1 << num);
                
                if (solve(board, rows, cols, boxes)) {
                    return true;
                }
                
                // Backtrack
                board[row][col] = '.';
                rows[row] &= ~(1 << num);
                cols[col] &= ~(1 << num);
                boxes[boxIdx] &= ~(1 << num);
            }
        }
        
        return false;
    }
};

















// class Solution {
//     public:
//         void solveSudoku(vector<vector<char>>& board) {
//             solve(board);
//         }
    
//     private:
//         bool check(int i, int j, char c, vector<vector<char>>& board) {
//             for (int k = 0; k < 9; k++) {
//                 if (board[i][k] == c) return false;
//                 if (board[k][j] == c) return false;
//                 if (board[3*(i/3)+k/3][3*(j/3)+k%3] == c) return false;
//             }
//             return true;        
//         }
//         bool solve(vector<vector<char>>& board) {
//             for (int i = 0; i < 9; i++) {
//                 for (int j = 0; j < 9; j++) {
//                     if (board[i][j] == '.') {
//                         for (char c = '1'; c <= '9'; c++) {
//                             if (check(i, j, c, board)) {
//                                 board[i][j] = c;
//                                 if (solve(board)) return true;
//                                 board[i][j] = '.';
//                             }
//                         }
//                         return false;
//                     }
                    
//                 }
//             }
//             return true;
//         }
//     };
    